Update security.md (#20981)
grafana/grafana: docs/sources/installation/security.md

# Security

> If you run non-Grafana web services on your Grafana server or within its local network, then they might be vulnerable to exploitation through the Grafana data source proxy.

Reviewer: `Data source proxy and protecting internal services` and `Viewer query permissions` are explaining two different things, though to some extent related but `Viewer query permissions` applies to both data source proxy requests and backend data source queries. 

Data source proxy calls: https://grafana.com/docs/grafana/latest/http_api/data_source/#data-source-proxy-calls
Data source backend query seems to be lacking from HTTP API, but should be `/api/tsdb/query`.

With your changes all sections is in H2 with intro text only referring to "exploitation through the Grafana data source proxy" which seems incorrect to me. I would consider having them in two different H2 headers or re-write the intro-text. 

Thoughts?

Author: I found that endpoint mentioned here: https://grafana.com/docs/grafana/v6.2/plugins/developing/backend-plugins-guide/#request-format

If something needs to be added to the API section, please put in a PR, even if it is sparse.

I am unclear on what relating to the tsdb query API that you cite should be added here.

I rewrote the introduction, please review and let me know if you have additional suggestions.

To prevent this type of exploitation from happening, we recommend that you apply one or more of the precautions listed below.

## Limit IP addresses/hostnames for data source URL

You can configure Grafana to only allow certain IP addresses or hostnames to be used as data source URL and by that proxied through the Grafana data source proxy. Refer to [data_source_proxy_whitelist]({{< relref "configuration/#data-source-proxy-whitelist" >}}) for usage instructions.

## Firewall rules

> Configure a firewall, for example using iptables, to restrict Grafana from making network requests to certain internal web services.

Author: Can we expand on this? Are there resources or guidelines we can cite or link to?

Reviewer: Guess we didn't want to have detailed instructions here since you can probably do this in 100 different ways using 100 different tools, but I guess a link to https://en.wikipedia.org/wiki/Iptables or some other would be nice.

## Proxy server

> Require all network requests being made by Grafana to go through a proxy server.

Author: Can we expand on this? Are there resources or guidelines we can cite or link to?

Reviewer: Maybe https://grafana.com/docs/grafana/latest/installation/behind_proxy/

## Viewer query permissions

> Users with the Viewer role can still issue any possible query to all data sources available in the **organization**, not just the queries that are defined on the dashboards the user with Viewer role has permissions to view.

Reviewer: End of this sentence (after comma) feels weird to me, but maybe me not used to reading/writing in this way?

Author: No, it is a terrible sentence. I am just not sure what it is trying to say, so not sure how to rewrite it.

Reviewer 2: It is saying that a person with the viewer role can issue any query they want (if they have some IT skills). An admin might think that viewers can only issue queries that are defined in the dashboards but that would be a false sense of security.

For example, if there is a Grafana instance with one data source, one dashboard and one panel which has one query defined then you might think that is all a viewer can see. But a viewer has access to send queries to the data source so with a command-line tool like curl (there are lots of tools for this), the viewer can make their own query and send it to the data source and potentially access sensitive data.

Author: Updated

There are a couple of ways you can restrict data source query access:

> - Create multiple data sources with some restrictions added in data source config that restrict access (like database name or credentials). Then use the [Data Source Permissions]({{< relref "../permissions/datasource_permissions.md" >}}) Enterprise feature to restrict user access to the data source in Grafana.

Author: @DanCech, is this an argument in favor of the Enterprise version?

Reviewer 3: yes

# Security

If you run non-Grafana web services on your Grafana server or within its local network, then they might be vulnerable to exploitation through the Grafana data source proxy or other methods.

To prevent this type of exploitation from happening, we recommend that you apply one or more of the precautions listed below.

## Limit IP addresses/hostnames for data source URL

> You can configure Grafana to only allow certain IP addresses or hostnames to be used as data source URL and by that proxied through the Grafana data source proxy. Refer to [data_source_proxy_whitelist]({{< relref "configuration/#data-source-proxy-whitelist" >}}) for usage instructions.

Reviewer 4: ```suggestion
You can configure Grafana to only allow certain IP addresses or hostnames to be used as data source URLs and proxied through the Grafana data source proxy. Refer to [data_source_proxy_whitelist]({{< relref "configuration/#data-source-proxy-whitelist" >}}) for usage instructions.
```

## Firewall rules

> Configure a firewall to restrict Grafana from making network requests to certain internal web services. 

Reviewer 2: Maybe replace `certain` with `sensitive` or `restricted`

There are many firewall tools available, refer to the documentation for your specific security tool. For example, Linux users can use [iptables](https://en.wikipedia.org/wiki/Iptables).

## Proxy server

> Require all network requests being made by Grafana to go through a proxy server. Refer to [Running Grafana behind a reverse proxy](https://grafana.com/docs/grafana/latest/installation/behind_proxy/) for examples.

Reviewer 4: This section is confusing using an outbound proxy for requests *from* Grafana (used for outbound requests from the datasource and plugin proxies, etc) with using a reverse proxy in front of grafana (which processes client requests).

Author: What needs to change in this passage?

Reviewer 5: @oddlittlebird -- remove the sentance "Refer to [Running Grafana behind a reverse proxy](https://grafana.com/docs/grafana/latest/installation/behind_proxy/) for examples"  

Note -- that doc shows how to proxy requests going **to** grafana, not requests coming **from** grafana.

Author: Done.
